runtime/registry.go

```diff
@@ -502,3 +503,54 @@ func (r *registryCache) ensureProjectParser(ctx context.Context, instanceID stri
 		r.logger.Error("could not create project parser", zap.Error(err), zap.String("instance_id", instanceID), observability.ZapCtx(ctx))
 	}
 }
+
+// instanceHeartbeatEmitter emits instance heartbeat events periodically after specified duration until closed.
+type instanceHeartbeatEmitter struct {
+	rt       *Runtime
+	duration time.Duration
+
+	closeChan chan struct{}
+}
+
+func newInstanceHeartbeatEmitter(rt *Runtime, duration time.Duration) *instanceHeartbeatEmitter {
+	return &instanceHeartbeatEmitter{
+		rt:       rt,
+		duration: duration,
+	}
+}
+
+func (i *instanceHeartbeatEmitter) emit() {
+	ticker := time.NewTicker(i.duration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			instances, err := i.rt.registryCache.list()
+			if err != nil {
+				i.rt.logger.Warn("failed to send instance heartbeat event, instance listing failed", zap.Error(err))
+				return
```

Comment:
Return here is weird since the error doesn't cause a crash. So it might look like it recovered. Probably better to `continue` and let it keep printing error messages once per minute.